crates/rover-client/src/introspection/schema.rs

```diff
@@ -0,0 +1,237 @@
+use sdl_encoder::{
+    Directive, EnumDef, EnumValue, Field, FieldValue, InputObjectDef, InputValue, InterfaceDef,
+    ObjectDef, ScalarDef, Schema as SDL, UnionDef,
+};
+use serde::Deserialize;
+use std::convert::TryFrom;
+
+pub type Introspection = introspect::introspection_query::ResponseData;
+pub type SchemaTypes = introspect::introspection_query::IntrospectionQuerySchemaTypes;
+pub type SchemaDirectives = introspect::introspection_query::IntrospectionQuerySchemaDirectives;
+pub type FullTypeFields = introspect::introspection_query::FullTypeFields;
+pub type FullTypeFieldArgs = introspect::introspection_query::FullTypeFieldsArgs;
+pub type __TypeKind = introspect::introspection_query::__TypeKind;
+
+// Represents GraphQL types we will not be encoding to SDL.
+const GRAPHQL_NAMED_TYPES: [&str; 13] = [
+    "__Schema",
+    "__Type",
+    "__TypeKind",
+    "__Field",
+    "__InputValue",
+    "__EnumValue",
+    "__DirectiveLocation",
+    "__Directive",
+    "Boolean",
+    "String",
+    "Upload",
+    "Int",
+    "ID",
+];
+
+// Represents GraphQL directives we will not be encoding to SDL.
+const GRAPHQL_DIRECTIVES: [&str; 3] = ["skip", "include", "deprecated"];
+
+/// A representation of a GraphQL Schema.
+///
+/// Contains Schema Types and Directives.
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Schema {
+    types: Vec<SchemaTypes>,
+    directives: Vec<SchemaDirectives>,
+}
+
+impl Schema {
+    /// Encode Schema into an SDL.
+    pub fn encode(self) -> String {
+        let mut sdl = SDL::new();
+
+        // Exclude GraphQL directives like 'skip' and 'include' before encoding directives.
+        self.directives
+            .into_iter()
+            .filter(|directive| !GRAPHQL_DIRECTIVES.contains(&directive.name.as_str()))
+            .for_each(|directive| Self::encode_directives(directive, &mut sdl));
+
+        // Exclude GraphQL named types like __Schema before encoding full type.
+        self.types
+            .into_iter()
+            .filter(|type_| match type_.full_type.name.as_deref() {
+                Some(name) => !GRAPHQL_NAMED_TYPES.contains(&name),
+                None => false,
+            })
+            .for_each(|type_| Self::encode_full_type(type_, &mut sdl));
+
+        sdl.finish()
+    }
+
+    fn encode_directives(directive: SchemaDirectives, sdl: &mut SDL) {
+        let mut directive_ = Directive::new(directive.name);
+        directive_.description(directive.description);
+        for location in directive.locations {
+            // Location is of a __DirectiveLocation enum that doesn't implement
+            // Display (meaning we can't just do .to_string). This next line
+            // just forces it into a String with format! debug.
+            directive_.location(format!("{:?}", location));
+        }
+
+        sdl.directive(directive_)
+    }
+
+    fn encode_full_type(type_: SchemaTypes, sdl: &mut SDL) {
+        let ty = type_.full_type;
```

Comment:
I assume we're using `ty` here because `type` is a reserved word? Is that a common abbreviation in Rust?

Reply:
`type` is a reserved word, but I usually use `ty` as it's a common compiler abbreviation. Two examples to check out here:
- the rust compiler has a dedicated [ty module](https://github.com/rust-lang/rust/tree/master/compiler/rustc_middle/src/ty), and
- wasm-tools, specifically the wasm-encoder, uses [ty in most of their encodings](https://github.com/bytecodealliance/wasm-tools/blob/a1535a01ea425312258729a277a2fcee8e98aadb/crates/wasm-smith/src/encode.rs#L50-L81)